MiniDB/DataBase.cs

```diff
@@ -0,0 +1,1170 @@
+                return undos_count > 0;
+            }
+        }
+
+        /// <summary>
+        /// Gets the current DB Version
+        /// - Any DB newer than this value is considered too new to handle.
+        /// - has internal set method so that Serializer can access it.
+        /// </summary>
+        [JsonProperty]
+        public float DBVersion { get; internal set; }
+
+        /// <summary>
+        /// Gets the filename/path that is used to cache the collection in
+        /// </summary>
+        protected string Filename { get; private set; }
+
+        /// <summary>
+        /// Gets the method that is used to migrate old versions of the db
+        /// </summary>
+        protected Func<DBMigrationParameters, JToken> Migrator { get; }
+
+        /// <summary>
+        /// Gets Json serialized value of this as string
+        /// </summary>
+        protected string SerializeData
+        {
+            get
+            {
+                lock (Locker)
+                {
+                    // TODO: compress https://dotnet-snippets.de/snippet/strings-komprimieren-und-dekomprimieren/1058
+                    return JsonConvert.SerializeObject(this, new DataBaseSerializer<T>());
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the databse for caching transactions for undo/redo.
+        /// </summary>
+        private DataBase<DBTransaction<T>> Transactions_DB { get; set; }
+        #endregion
+
+        #region public methods   
+        /// <summary>
+        /// Helper method determining if there is a conflict between changes to two different db's based on both db's changing the same item (determined by ID).
+        /// </summary>
+        /// <param name="db1">first db</param>
+        /// <param name="db2">second db</param>
+        /// <returns>true if conflict</returns>
+        public static bool ConflictIfSameItemChangedById(DataBase<T> db1, DataBase<T> db2)
+        {
+            bool result = false;
+
+            // any transaction in db1 has the same item db as any tranasction in db2
+            result = db1.Transactions_DB.Any(x => db2.Transactions_DB.Any(y => x.Item_ID == y.Item_ID));
+            return result;
+        }
+
+        /// <summary>
+        /// Clear the mutex when a using statement leaves scope (or destructor is called)
+        /// </summary>
+        public void Dispose()
+        {
+            if (this.mut != null && this.mut.WaitOne(TimeSpan.FromSeconds(5), false))
+            {
+                this.mut.ReleaseMutex();
+                this.mut.Close();
+                this.mut = null;
+            }
+        }
+
+        /// <summary>
+        /// Undo last undoable transaction and raise property changed events on changed item.
+        /// TODO: make sure to use locker everywhere that it is messing with the data
+        /// </summary>
+        public void Undo()
```

Comment:
The Undo() and Redo() methods are far too long. Refactor them into shorter methods. For example, every time you feel the need to add a comment saying what the code is doing, that comment should probably be a method name.

Comments should say why you're doing something, not what you're doing. If you need a comment to explain what you're doing, then your code isn't clear enough yet.